THE-app/app/(tabs)/booksScreen.tsx

```diff
@@ -56,18 +57,91 @@ function BooksScreen() {
     return data?.viewer?.books?.hits?.filter((book: Book) => book.valid) || [];
   }, [data]);
 
+  const [filterText, setFilterText] = useState('');
+  const [selectedSubject, setSelectedSubject] = useState('');
+  const [selectedLevel, setSelectedLevel] = useState('');
+
+  const [subjectOptions, levelOptions] = useMemo(() => {
+    const subjectsSet = new Set<string>();
+    const levelsSet = new Set<string>();
+
+    if (data?.viewer?.books?.hits) {
+      data.viewer.books.hits.forEach((book: Book) => {
+        book.subjects.forEach((subject) => subjectsSet.add(subject.name));
+        book.levels.forEach((level) => levelsSet.add(level.name));
+      });
+    }
+
+    const subjectsArray = Array.from(subjectsSet).sort();
+    const levelsArray = Array.from(levelsSet).sort();
+
+    return [
+      subjectsArray.map(subject => <Select.Item key={subject} label={subject} value={subject} />),
+      levelsArray.map(level => <Select.Item key={level} label={level} value={level} />)
+    ];
+  }, [data]);
+
+  const filteredBooks = useMemo(() => {
+    const searchText = filterText.toLowerCase();
+
+    return books
+      .filter((book) => {
+        const matchesText = !filterText || Object.values(book).some(value =>
+          typeof value === 'string' && value.toLowerCase().includes(searchText)
+        );
+        const matchesSubject = !selectedSubject || book.subjects.some(subject => subject.name === selectedSubject);
+        const matchesLevel = !selectedLevel || book.levels.some(level => level.name === selectedLevel);
+        return matchesText && matchesSubject && matchesLevel;
```

Comment:
This is nice for debugging, but because you're storing the partial conditions in variables, you're disabling the short-circuit system.

For example, if `matchesText` is false, it's not necessary to compute `matchesSubject` or `matchesLevel`.